avocado/utils/pci.py

```diff
@@ -0,0 +1,310 @@
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+#
+# See LICENSE for more details.
+#
+# Copyright: 2016 IBM
+# Author: Narasimhan V <sim@linux.vnet.ibm.com>
+#
+# Author: Kleber Sacilotto de Souza <klebers@linux.vnet.ibm.com>
+# Author: Daniel Kreling <kreling@linux.vnet.ibm.com>
+# for get_memory_address() and get_mask()
+
+"""
+Module for all PCI devices related functions.
+"""
+
+
+import re
+import os
+from . import process
+
+
+def get_domains():
+    """
+    Gets all pci domains.
+
+    :return: List of pci domains.
+    """
+    cmd = "lspci -D"
+    output = process.system_output(cmd, shell=True, ignore_status=True)
+    if output:
+        domains = []
+        for line in output.splitlines():
+            domains.append(line.split(":")[0])
+        return list(set(domains))
+
+
+def get_num_devices_in_domain(domain):
+    """
+    Gets number of devices in a pci domain.
+
+    :parm domain: pci domain.
+
+    :return: number of devices in a pci domain.
+    """
+    cmd = "ls -l /sys/class/*/ -1"
+    output = process.system_output(cmd, shell=True, ignore_status=True)
+    if output:
+        count = 0
+        for line in output.splitlines():
+            if domain in line:
+                count += 1
+        return count
+
+
+def get_disks_in_device(device):
+    """
+    Gets disks in a pci device.
+
+    :parm device: pci device.
+
+    :return: list of disks in a pci device.
+    """
+    disks_path = "/dev/disk/by-path/"
+    devices_list = []
+    for dev in os.listdir(disks_path):
+        if device in dev:
+            link = os.readlink("%s%s" % (disks_path, dev))
+            devices_list.append(os.path.abspath("%s%s" % (disks_path, link)))
+    return devices_list
+
+
+def get_interfaces_in_device(device):
+    """
+    Gets ethernet interface in a pci device.
+
+    :parm device: pci device.
+
+    :return: list of interfaces in a pci device.
+    """
+    iface_path = "/sys/class/net/"
+    interfaces_list = []
+    for iface in os.listdir(iface_path):
+        if device in os.readlink("%s%s" % (iface_path, iface)):
+            interfaces_list.append(iface)
+    return interfaces_list
+
+
+def get_pci_fun_list(device):
+    """
+    Gets list of functions in the given pci device.
+    Example: in device 0000:03:00, functions are 0000:03:00.0 and 0000:03:00.1
+
+    :parm device: pci device.
+
+    :return: list of functions in a pci device.
+    """
+    return list(dev for dev in get_pci_devices() if device in dev)
+
+
+def get_slot_from_sysfs(device):
+    """
+    Gets the pci slot of given device.
+    Specific for ppc64 processor
+
+    :parm device: pci device.
+
+    :return: slot of pci device from sysfs.
+    """
+    if not os.path.isdir('/sys/bus/pci/devices/%s' % device):
+        return
+    cmd = "cat /sys/bus/pci/devices/%s/devspec" % device
+    devspec = process.system_output(cmd, shell=True).strip('\n')
+    cmd = "cat /proc/device-tree/%s/ibm,loc-code" % devspec
```

Comment:
```
if not os.path.isdir('/sys/bus/pci/devices/%s' % device):
    return
```
This will take care when such errors might occur.
Do you have a case where it does not ?